### PR TITLE
Fix broken tests on elements and various cleanups

### PR DESCRIPTION
**Pull request by cdecker**
This was supposed to be a regular cleanup PR, but given that a couple of tests broke along the way for `liquid-regtest`, it took me a bit longer.

 - Create test directory: some tests, especially those testing tools, will not create the working directory themselves, so we create it when the fixture is created.
 - Strengthened a couple of tx creation tests, mainly minimizing the array allocation sizes and ensuring everything is set up correctly (subsumes chain specific modifications).
 - MIgrate away from deprecate `pylightning` dependency towards `pyln-client`
 - There is a lingering test failure in which we fail to remove the working directory. I added a printing of the file listing to see what files are causing that issue. It then rethrows the error so we don't just swallow it.

There is one more thing I need to find before un-drafting:

 - [x] If running with `DEVELOPER=1`, `TEST_DB_PROVIDER=postgres` and `VALGRIND=1` there is a memory leak when retrieving UTXOs for `txprepare`.

Changelog-None

### CODE REVIEW

**Comment by cdecker** [comment]
I'll have to revisit the memleak when running with `postgres` later, don't want to hold up the other cleanups since they are rather nice imho.

**Comment by cdecker** [comment]
Managed to fix two more regressions:

 - [X] The memleak that was present in `postgres` due to unordered UTXOs was also a leak with `sqlite3` but manifested only rarely. Now fixed.
 - [ ] A new test that broke because it expected mainnet fees on liquid. Parameterized using `chainparams`

This should now be complete and ready to be reviewed. Paging release captain @niftynei :wink:

**Comment by cdecker** [comment]
Here are the results from running all the various configurations:

```
[{"['TEST_NETWORK=regtest', 'VALGRIND=0', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=postgres', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=regtest', 'VALGRIND=0', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=sqlite3', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=regtest', 'VALGRIND=1', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=postgres', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=regtest', 'VALGRIND=1', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=sqlite3', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=liquid-regtest', 'VALGRIND=0', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=postgres', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=liquid-regtest', 'VALGRIND=0', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=sqlite3', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=liquid-regtest', 'VALGRIND=1', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=postgres', 'DEVELOPER=1']": '0'},
 {"['TEST_NETWORK=liquid-regtest', 'VALGRIND=1', 'PYTEST_PAR=10', 'TEST_DB_PROVIDER=sqlite3', 'DEVELOPER=1']": '0'}]
```

**Comment by rustyrussell** [comment]
Simple rebase: test conflicts.

File: Makefile [hunk 1]
```diff
@@ -272,7 +272,7 @@ ifeq ($(PYTEST),)
 	exit 1
 else
 # Explicitly hand DEVELOPER and VALGRIND so you can override on make cmd line.
-	PYTHONPATH=`pwd`/contrib/pyln-client:`pwd`/contrib/pyln-testing:`pwd`/contrib/pylightning:`pwd`/contrib/pyln-proto/:$(PYTHONPATH) TEST_DEBUG=1 DEVELOPER=$(DEVELOPER) VALGRIND=$(VALGRIND) $(PYTEST) tests/ $(PYTEST_OPTS)
+	PYTHONPATH=`pwd`/contrib/pyln-client:`pwd`/contrib/pyln-testing:`pwd`/contrib/pyln-proto/:$(PYTHONPATH) TEST_DEBUG=1 DEVELOPER=$(DEVELOPER) VALGRIND=$(VALGRIND) $(PYTEST) tests/ $(PYTEST_OPTS)
```
Comment (darosior) [inline Makefile]: :tada:

File: bitcoin/tx.c [hunk 2]
```diff
@@ -98,28 +98,43 @@ static struct amount_sat bitcoin_tx_compute_fee(const struct bitcoin_tx *tx)
...
-			assert(pos == -1);
-			pos = i;
-		}
+	for (pos = 0; pos < tx->wtx->num_outputs; pos++) {
+		if (elements_tx_output_is_fee(tx, pos))
+			break;
 	}
 
-	if (pos == -1) {
+	if (pos >= tx->wtx->num_outputs) {
```
Comment (darosior) [inline bitcoin/tx.c]: It cannot be `>`, can it ?

File: bitcoin/tx.c [hunk 3]
```diff
@@ -98,28 +98,43 @@ static struct amount_sat bitcoin_tx_compute_fee(const struct bitcoin_tx *tx)
...
 		w = tal(tx->output_witscripts, struct witscript);
 		w->ptr = tal_arr(w, u8, 0);
-		tx->output_witscripts[tx->wtx->num_outputs] = w;
+
+		/* Make sure we have a place to stash the witness script in. */
+		if (tal_count(tx->output_witscripts) < pos + 1) {
+			fprintf(stderr, "XXX resizing array to %d to fit item at %d\n", pos + 1, pos);
```
Comment (darosior) [inline bitcoin/tx.c]: Trailing debug line ?

File: tests/test_closing.py [hunk 4]
```diff
@@ -465,7 +465,7 @@ def test_closing_fee(node_factory, bitcoind, chainparams):
         'funder_feerate_per_kw': 29006,
         'fundee_feerate_per_kw': 27625,
         'close_initiated_by': 'funder',
-        'expected_close_fee': 20333
+        'expected_close_fee': 33533 if chainparams['elements'] else 20333
```
Comment (darosior) [inline tests/test_closing.py]: How is this calculated ? Asking for ~~a friend~~ #3592 where I wondered how much to bump it in functional tests (e.g. https://github.com/ElementsProject/lightning/pull/3592/commits/616a432bd03359afc3abdecb5bbe33b8ced3ccda#diff-d03575dee065f2dc319cf0e8d39ebe95R548)

File: wallet/wallet.c [hunk 5]
```diff
@@ -403,8 +403,10 @@ static const struct utxo **wallet_select(const tal_t *ctx, struct wallet *w,
 		 * confirmation height and that it is below the required
 		 * maxheight (current_height - minconf) */
 		if (maxheight != 0 &&
-		    (!u->blockheight || *u->blockheight > maxheight))
+		    (!u->blockheight || *u->blockheight > maxheight)) {
+			tal_free(u);
```
Comment (darosior) [inline wallet/wallet.c]: Doing so will unreserve the utxo, right ? If so, I think that's what I missed to fix #3591 in #3593 !

File: bitcoin/tx.c (hunk 2 above)
Comment (cdecker) [reply bitcoin/tx.c]: You're right. For some reason I'm in the habit of being overly cautious in these cases (spurious increments somewhere), but I'll fixup in this case :-)

File: tests/test_closing.py (hunk 4 above)
Comment (cdecker) [reply tests/test_closing.py]: I usually just binary search until I find params that work on both sides (keeping as many parameters as possible fixed).

I have not found a good way to derive from first principles, other than reimplementing the logic in python, which'd just double the maintenance effort without a major upside.

File: wallet/wallet.c (hunk 5 above)
Comment (cdecker) [reply wallet/wallet.c]: Not yet: we are in `wallet_select` so the coins have not been marked as selected yet (that happens 6 lines down). It's the destructor added to the tallocated array on line 361 that'd free them, but they were not being added yet (just skipped) before this fix.

Comment (cdecker) [reply wallet/wallet.c]: That's a roundabout way of saying that to unmark selected UTXOs you need to `tal_free` the `utxos` array :-)

**Comment by rustyrussell** [comment]
Ack c85988f2da83518d09a1575ab21a0dab83f95ea8

**Pull request merged**